Implement initial GA4GH DRS Support.

jmchilton opened the pull request:
Builds on #13947.

- Implement APIs for serving Galaxy's public datasets as DRS objects.
- Implement support in ``galaxy.files.uris`` for DRS URIs so that upload, etc... can use DRS URIs from other data repositories.

Together these represent effectively both client and server or producer and consumer support for the DRS API.

The ways this implementation can be improved/extended are endless:

- New object ID patterns can be implemented to allow access to composite data, metadata files from a bundle view of datasets.
- New object ID patterns can be implemented to allow access to collections as bundles of datasets.
- More auth types and AccessMethods could be implemented for loading datasets from clients.
- Some sort of scheme for providing authenticated URLs to non-public data could be implemented to provide access to more of Galaxy's data if user's allow/enable/etc....

Despite the limitations - I think having the guts in like this provide useful points for interested parties to go and add functionality to enable their desired use cases.

## How to test the changes?
(Select all options that apply)
- [X] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

bgruening commented on doc/source/admin/galaxy_options.rst:
```diff
@@ -3756,6 +3756,43 @@
+:Description:
+    Service ID for GA4GH services (exposed via the service-info
+    endpoint for the Galaxy DRS API). If unset, one will be generated
+    using the URL the target API requests are made against.
```
I guess an example for usegalaxy.org would be nice to add and since those IDs should be stable maybe we should make this explicit? Maybe Galaxy servers are accessed via different URLs, e.g. subdomains ...

bgruening commented on lib/galaxy/webapps/galaxy/services/datasets.py:
```diff
@@ -367,6 +375,50 @@ def show_inheritance_chain(
+            checksums.append(Checksum(type=type, checksum=checksum))
+
+        if len(checksums) == 0:
+            # fire off a checksuming task and return a 202?
```
I guess so. This will happen a lot, since we do not have checksum for all datasets by default, do we?

mvdbeek replied:
Are there some instructions on how this case should be handled ? Should we ask this upstream ?

jmchilton replied:
The response codes sections explicitly calls out 202s... it seems the "correct" approach to me. https://ga4gh.github.io/data-repository-service-schemas/preview/release/drs-1.0.0/docs/#_responses

jmchilton replied:
Rebased this on top of https://github.com/galaxyproject/galaxy/pull/14576 and it now fires off a Celery request to MD5 the dataset and returns a 202. I've updated the DRS client to handle the 202 responses.

mvdbeek commented on lib/galaxy/webapps/galaxy/api/drs.py:
```diff
@@ -0,0 +1,113 @@
+                return FileResponse(file_name, headers=headers)
+        elif isinstance(display_data, ZipstreamWrapper):
+            return StreamingResponse(display_data.response(), headers=headers)
+        return StreamingResponse(display_data, headers=headers)
```
```suggestion
        display_data, headers = self.service.display(trans, object_id, False, None, None, True)
        assert isinstance(display_data, IOBase)
        return FileResponse(file_name, headers=headers)
```

mvdbeek commented on lib/galaxy/webapps/galaxy/api/drs.py:
```diff
@@ -0,0 +1,113 @@
+
+    @router.get(
+        "/api/drs_download/{object_id}",
+        response_class=StreamingResponse,
```
```suggestion
        response_class=FileResponse,
```

mvdbeek commented on lib/galaxy/webapps/galaxy/api/drs.py:
```diff
@@ -0,0 +1,113 @@
+from starlette.responses import (
+    FileResponse,
+    StreamingResponse,
+)
```
```suggestion
from starlette.responses import FileResponse
```

mvdbeek commented on lib/galaxy/webapps/galaxy/services/datasets.py:
```diff
@@ -367,6 +375,50 @@ def show_inheritance_chain(
+
+        if len(checksums) == 0:
+            # fire off a checksuming task and return a 202?
+            checksums.append(Checksum(type="md5", checksum="dummysum"))
```
maybe raise an exception for now ?

jmchilton replied:
Rebased this on top of https://github.com/galaxyproject/galaxy/pull/14576 and it now fires off a Celery request to MD5 the dataset and returns a 202. I've updated the DRS client to handle the 202 responses.

mvdbeek commented on lib/galaxy/schema/drs/gen.sh:
```diff
@@ -0,0 +1,10 @@
+# https://github.com/koxudaxi/datamodel-code-generator
+for model in AccessMethod Checksum DrsObject Error AccessURL ContentsObject DrsService
+do
+    datamodel-codegen --url "https://raw.githubusercontent.com/ga4gh/data-repository-service-schemas/master/openapi/components/schemas/${model}.yaml" --output "$model.py"
```
Can you use a permalink, e.g a tag or a commit in urls ?

jmchilton replied:
I mean ... we wouldn't want to regenerate this if nothing changed 😆.

mvdbeek commented on lib/galaxy/webapps/galaxy/services/datasets.py:
```diff
@@ -367,6 +375,50 @@ def show_inheritance_chain(
+    def get_drs_object(
+        self,
+        trans: ProvidesHistoryContext,
+        object_id: EncodedDatabaseIdField,
```
Since I suppose there'll be some pressure for the object_id to be stable: Should we use something like `hda:uuid` instead of the encoded database id ? That way we can serve objects that are not hdas in the future, and we avoid issues if someone rotates the id_secret

jmchilton replied:
There really should be no such pressure in my opinion - the purpose of DRS is data transfer for workflow and task execution systems.

> The primary functionality of DRS is to map a logical ID to a means for physically retrieving the data represented by the ID. The sections below describe the characteristics of those IDs, the types of data supported, and how the mapping works.

There should be no expectation that these are permanent IMO.

That said... I'm open to requiring the "hda:" prefix up front... and with encoding the IDs with a different key that can be more stable (like we do with the job files API - but in that case we do it for added security).

mvdbeek reviewed (commented):
Looks pretty cool!

mvdbeek commented on lib/galaxy/webapps/galaxy/api/drs.py:
```diff
@@ -0,0 +1,105 @@
+    ):
+        display_data, headers = self.service.display(trans, object_id, False, None, None, True)
+        assert isinstance(display_data, IOBase)
+        return FileResponse(file_name, headers=headers)
```
I'm sorry for the bad suggestion, should have been:
```suggestion
        return FileResponse(getattr(display_data, "name"), headers=headers)
```
The getattr prevents mypy from complaining, there should always be a `.name` attribute.

uwwint committed:
Implement access_id method to resolve object with expensive url generation schema. See https://ga4gh.github.io/data-repository-service-schemas/preview/develop/docs/#tag/AccessMethodModel

nuwang commented on lib/galaxy/files/uris.py:
```diff
@@ -51,6 +52,10 @@ def stream_url_to_file(
+    elif path.startswith("drs://"):
+        with tempfile.NamedTemporaryFile(prefix=prefix, delete=False) as temp:
+            temp_name = temp.name
+            fetch_drs_to_file(path, temp_name)
```
@jmchilton Can this be implemented as a file source instead?  Granted, the file source's list method would return just the initial `source_path` itself, unless it's a collection/bundle, in which case it could return the bundled objects? I ask because as a filesource, it would be far easier to use existing functionality to read auth off the profile, vault etc?

jmchilton replied:
I started with the same idea... but to me the abstraction doesn't work because nothing roots ```drs://``` - none of the UI navigation tools would make sense without that? This implementation seems to make more sense - every place we can fetch file sources uses this abstraction anyway so all the cool things we can do with file sources we can do here.

There is some chance I'm being dense about this though - if that would enable vault integration in someway I don't understand maybe sketch a PR or describe the use case and implementation in more detail?

nuwang replied:
The main issue I was wondering about was how auth would be handled with DRS? Specifically, my understanding was that a DRS server may require a [bearer token to gain access](https://ga4gh.github.io/data-repository-service-schemas/preview/release/drs-1.0.0/docs/#_making_drs_requests). How would that token be injected here? https://github.com/galaxyproject/galaxy/pull/13949/files#diff-394e589bbc6b016c487d5373c64a4376c6ded4ebb7633d05487a42ca5ac8e218R38

With a filesource, the mechanisms for reading a token from the user profile, vault etc. already exist. Would we have to duplicate that functionality here?

At the risk of stretching the file source abstraction too far, what if all uris were delegated to filesources, including http, base64 and drs? We could add a `supports_root` method to a file source, indicating that it does not support a root listing and can therefore be excluded from the UI?

jmchilton replied:
Given the lack of the listing (ls) functionality for DRS, http, etc... - I don't think I'm for this change. Maybe like a parent plugin framework that can handle either file sources or URI fetchers maybe? I think the fact that we've synchronized all URI fetching through this function makes that feasible. 

If you've got a particular test case / use case where that really cleans up things and enables a new DRS source to be used though... I certainly think it is worth considering an implementation.

nuwang replied:
@jmchilton I had a discussion about this with @uwwint and we have a proposal for how file sources can be unified with URI fetchers. I'll create a separate issue to track that.

jmchilton replied:
@nuwang Awesome - I look forward to seeing what y'all come up with.

nuwang reviewed (commented):
Not sure what remains to be done in this PR, but I wonder whether this would have been better off as two PRs, one to fetch data from DRS, and another to expose Galaxy data through DRS? The former at least, appears to be complete?

bgruening commented on lib/galaxy/util/drs.py:
```diff
@@ -79,7 +79,16 @@ def fetch_drs_to_file(
+        access_url = access_method["access_url"]
+    except KeyError:
+        access_id = access_method["access_id"]
+        access_get_url = f"{scheme}://{netspec}/ga4gh/drs/v1/objects/{object_id}/access/{access_id}"
```
Should this (re-)use `get_url` from above?

bgruening commented on doc/source/admin/galaxy_options.rst:
```diff
@@ -3823,6 +3823,43 @@
+:Description:
+    Service ID for GA4GH services (exposed via the service-info
+    endpoint for the Galaxy DRS API). If unset, one will be generated
+    using the URL the target API requests are made against.
```
This should be `usegalaxy.org` for example I assume? I guess this is important to set, especially if you have subdomains.

bgruening replied:
Can we give here and maybe below a best-practice recommendation?

jmchilton replied:
I'm sure it isn't important to set at all 😆 but I've updated the documentation with links and my best guess at a recommendation for usegalaxy.org. Thanks for the attention to detail!

bgruening commented on doc/source/admin/galaxy_options.rst:
```diff
@@ -3823,6 +3823,43 @@
+:Description:
+    Service name for host organization (exposed via the service-info
+    endpoint for the Galaxy DRS API). If unset, one will be generated
+    using ga4gh_service_service_id.
```
```suggestion
    using ga4gh_service_id.
```

jmchilton replied:
Fixed, thank you!

jmchilton committed:
Flush out documentation and service-info for DRS API a bit.

jmchilton committed:
Some small tweaks to URL handling in DRS code.

jmchilton committed:
Updates for test file changes...

jmchilton committed:
Implement initial GA4GH DRS Support.

- Implement APIs for serving Galaxy's public datasets as DRS objects.
- Implement support in galaxy.files.uris for DRS URIs so that upload, etc... can use DRS URIs from other data repositories.

The ways this can be improved are endless:

- New object ID patterns can be implemented to allow access to composite data, metadata files from a bundle view of datasets.
- New object ID patterns can be implemented to allow access to collections as bundles of datasets.
- More auth types and AccessMethods could be implemented for loading datasets from clients.
- Some sort of scheme for providing authenticated URLs to non-public data could be implemented to provide access to more of Galaxy's data if user's can implement.

Despite all these limitations - I think having the guts in like this provide useful points for interested parties to go and add functionality to enable their desired usecases.

jmchilton committed:
Cleanup typing and dependencies around DRS.
